### PR TITLE
Delete BootNext as a post-reboot action

### PR DESCRIPTION
**Pull request by hughsie**
This should not be required, but then some firmware is broken.

Fixes https://github.com/fwupd/fwupd/issues/5938

Type of pull request:

- [ ] New plugin (Please include [new plugin checklist](https://github.com/fwupd/fwupd/wiki/New-plugin-checklist))
- [X] Code fix
- [ ] Feature
- [ ] Documentation

### CODE REVIEW

**Comment by hughsie** [comment]
@superm1 I'm not 100% on the vfunc name, other ideas very welcome.

**Review by superm1 — COMMENTED** [review]
why not just include it in the plugin startup routines?  Check if it's a different boot and just clean up there.

**Comment by hughsie** [comment]
> why not just include it in the plugin startup routines

I originally had this, but I thought it would be confusing -- if you're debugging boot stuff it's not 100% obvious that fwupd would be doing this *crazy thing*.

**Comment by superm1** [comment]
But don't we delete .CAP files the same way?

Imo the deleting behavior can be set to on by default and use a conf option to turn it off for debugging.

**Comment by hughsie** [comment]
> But don't we delete .CAP files the same way

Hmm, we do that using a glob in `fu_uefi_device_cleanup_esp` -- which we call in `klass_device->prepare` -- IIRC the reason we didn't do it at startup is that is required mounting the ESP on every boot -- which is slow. Are you suggesting we should move the .cap file deleting (and ESP mounting) to the same new vfunc so it only gets called after a successful update?

**Comment by superm1** [comment]
> > But don't we delete .CAP files the same way
> 
> Hmm, we do that using a glob in `fu_uefi_device_cleanup_esp` -- which we call in `klass_device->prepare` -- IIRC the reason we didn't do it at startup is that is required mounting the ESP on every boot -- which is slow. Are you suggesting we should move the .cap file deleting (and ESP mounting) to the same new vfunc so it only gets called after a successful update?

Well I think both should happen in the same place - and I think we should be doing both success or fail.

The only reason you need them around after a fail  is debugging, and we can do something like tie that to `fwupdate -L`

**Comment by hughsie** [comment]
> Well I think both should happen in the same place - and I think we should be doing both success or fail.

Something like ^^?

File: libfwupdplugin/fu-plugin.c [hunk 1]
```diff
@@ -1330,6 +1330,30 @@ fu_plugin_runner_reload(FuPlugin *self, FuDevice *device, GError **error)
...
+ * fu_plugin_runner_reboot_cleanup:
+ * @self: a #FuPlugin
+ * @device: a device
+ * @error: (nullable): optional return location for an error
+ *
+ * Performs cleanup actions after the reboot has been performed.
+ *
+ * Returns: #TRUE for success, #FALSE for failure
+ *
+ * Since: 1.9.5
```
Comment (superm1) [inline libfwupdplugin/fu-plugin.c:1343]: 1.9.6

File: libfwupdplugin/fu-plugin.h [hunk 2]
```diff
@@ -372,6 +372,17 @@ struct _FuPluginClass {
...
 	void (*to_string)(FuPlugin *self, guint idt, GString *str);
+	/**
+	 * reboot_cleanup:
+	 * @self: a #FuPlugin
+	 * @device: a device
+	 * @error: (nullable): optional return location for an error
+	 *
+	 * Performs cleanup actions after the reboot has been performed.
+	 *
+	 * Since: 1.9.5
```
Comment (superm1) [inline libfwupdplugin/fu-plugin.h:383]: 1.9.6

File: plugins/uefi-capsule/fu-uefi-capsule-plugin.c [hunk 3]
```diff
@@ -1146,6 +1147,67 @@ fu_uefi_capsule_plugin_coldplug(FuPlugin *plugin, FuProgress *progress, GError *
...
+	/* delete capsules */
+	if (!fu_uefi_capsule_plugin_cleanup_esp(self, error))
+		return FALSE;
+
+	/* delete any old variables */
+	if (!fu_efivar_delete_with_glob(FU_EFIVAR_GUID_FWUPDATE, "fwupd*-*", error))
+		return FALSE;
+
+	/* this should not be required, but, hey -- here were are */
+	if (fu_efivar_exists(FU_EFIVAR_GUID_EFI_GLOBAL, "BootNext")) {
```
Comment (superm1) [inline plugins/uefi-capsule/fu-uefi-capsule-plugin.c:1200]: Need to be careful here. Fwupd isn't the only thing that uses BootNext. I think it needs to only be deleted if the variable points to the Linux Firmware Updater entry.

File: libfwupdplugin/fu-plugin.c (hunk 1 above)
Comment (hughsie) [reply libfwupdplugin/fu-plugin.c:1343]: fixed.

File: libfwupdplugin/fu-plugin.h (hunk 2 above)
Comment (hughsie) [reply libfwupdplugin/fu-plugin.h:383]: fixed.

File: plugins/uefi-capsule/fu-uefi-capsule-plugin.c (hunk 3 above)
Comment (hughsie) [reply plugins/uefi-capsule/fu-uefi-capsule-plugin.c]: Agree; fixed.

**Review by superm1 — DISMISSED** [review]
Yeah, one minor nit and otherwise fine.

File: plugins/uefi-capsule/fu-uefi-capsule-plugin.c [hunk 4]
```diff
@@ -1146,6 +1147,113 @@ fu_uefi_capsule_plugin_coldplug(FuPlugin *plugin, FuProgress *progress, GError *
...
+	g_autoptr(GPtrArray) files = NULL;
+
+	/* delete any files matching the glob in the ESP */
+	esp_locker = fu_volume_locker(self->esp, error);
+	if (esp_locker == NULL)
+		return FALSE;
+	files = fu_path_get_files(esp_path, error);
+	if (files == NULL)
+		return FALSE;
+	pattern = g_build_filename(esp_path, "EFI/*/fw/fwupd*.cap", NULL);
```
[the remark this answers is not in the document]
Comment (hughsie) [reply plugins/uefi-capsule/fu-uefi-capsule-plugin.c:1169]: You mean `fu_uefi_get_esp_path_for_os()/fw/fwupd*.cap` perhaps?

[the remark this answers is not in the document]
Comment (hughsie) [reply plugins/uefi-capsule/fu-uefi-capsule-plugin.c]: Yup, that works -- fixed.

**Pull request merged**